sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala

```diff
@@ -146,13 +147,16 @@ case class CoalesceShufflePartitions(session: SparkSession) extends AQEShuffleRe
       Seq(collectShuffleStageInfos(r))
     case unary: UnaryExecNode => collectCoalesceGroups(unary.child)
     case union: UnionExec => union.children.flatMap(collectCoalesceGroups)
-    // If not all leaf nodes are exchange query stages, it's not safe to reduce the number of
-    // shuffle partitions, because we may break the assumption that all children of a spark plan
-    // have same number of output partitions.
+    case join: CartesianProductExec => join.children.flatMap(collectCoalesceGroups)
```

Comment:
The issue is that, if the plan itself does not require shuffle exchange, then we should not assume all of its leaf nodes are shuffle exchanges. How about adding a new pattern to make the non-unary plan which does not require shuffle into new groups. We can even remore the union pattern.
```
def doNotRequireShuffleExchange(plan) = plan.requiredChildDistribution.exists {
  case _: BroadcastDistribution | _: UnspecifiedDistribution => true
  case _ => false
}

case p if doNotRequireShuffleExchange(p) => p.children.flatMap(collectCoalesceGroups)
```